src/coreclr/jit/emit.cpp

```diff
@@ -9119,6 +9124,65 @@ void emitter::emitNxtIG(bool extend)
 #endif
 }
 
+//------------------------------------------------------------------------
+// emitRemoveLastInstruction: Remove the last instruction emitted; it has been optimized away by the
+// next instruction we are generating. `emitLastIns` must be non-null, meaning there is a
+// previous instruction. The previous instruction might have already been saved, or it might
+// be in the currently accumulating insGroup buffer.
+//
+// The `emitLastIns` is set to nullptr after this function. It is expected that a new instruction
+// will be immediately generated after this, which will set it again.
+//
+// NOTE: It is expected that the GC effect of the removed instruction will be handled by the newly
+// generated replacement(s).
+//
+void emitter::emitRemoveLastInstruction()
+{
+    assert(emitLastIns != nullptr);
+    assert(emitLastInsIG != nullptr);
+
+    JITDUMP("Removing saved instruction in %s:\n", emitLabelString(emitLastInsIG));
+    JITDUMPEXEC(dispIns(emitLastIns))
+
+    // We should assert it's not a jmp, as that would require updating the jump lists, e.g. emitCurIGjmpList.
+
+    BYTE* lastInsActualStartAddr = (BYTE*)emitLastIns - m_debugInfoSize;
+
+    if ((emitCurIGfreeBase <= (BYTE*)lastInsActualStartAddr) && ((BYTE*)lastInsActualStartAddr < emitCurIGfreeEndp))
+    {
+        // The last instruction is in the current buffer. That means the current IG is non-empty.
+        assert(emitCurIGnonEmpty());
+        assert((BYTE*)lastInsActualStartAddr < emitCurIGfreeNext);
+        assert(emitCurIGinsCnt >= 1);
+        assert(emitCurIGsize >= emitLastIns->idCodeSize());
+
+        size_t insSize    = emitCurIGfreeNext - (BYTE*)lastInsActualStartAddr;
+        emitCurIGfreeNext = (BYTE*)lastInsActualStartAddr;
+        emitCurIGinsCnt -= 1;
+        emitCurIGsize -= emitLastIns->idCodeSize();
+
+        // We're going to overwrite the memory; zero it.
+        memset(emitCurIGfreeNext, 0, insSize);
+    }
+    else
+    {
+        // The last instruction has already been saved. It must be the last instruction in the group.
+        assert((BYTE*)emitLastInsIG->igData + emitLastInsIG->igDataSize ==
+               (BYTE*)emitLastIns + emitSizeOfInsDsc(emitLastIns));
+        assert(emitLastInsIG->igInsCnt >= 1);
+        assert(emitLastInsIG->igSize >= emitLastIns->idCodeSize());
+
+        emitLastInsIG->igInsCnt -= 1;
+        emitLastInsIG->igSize -= (unsigned short)emitLastIns->idCodeSize();
```

Comment:
Save `emitLastIns->idCodeSize()` at the top of `if-else` and use it.

Reply:
Code changed, as requested.